### PR TITLE
Add option to supply custom matchers

### LINKED ISSUE #6: Honor the order of array of expected actions
Following the example presented in the [README](https://github.com/moxystudio/redux-mock-store-await-actions/blob/master/README.md), the `Promise` will resolve if `LOGIN_START` is dispatched either before or after `LOGIN_SUCCESS` as shown below:
```js
await waitForActions(store, ['LOGIN_START', 'LOGIN_SUCCESS']);

await waitForActions(store, ['LOGIN_SUCCESS', 'LOGIN_START']);
````
**Proposal:** the API shall accept a boolean option to specify whether the order of arrival of dispatched actions should match the order of the array of expected actions. The `Promise` rejects on the first mismatch.

### PR DESCRIPTION
**Pull request by acostalima**
BREAKING CHANGE: convert timeout argument to an options object. This commit introduces the matcher concept, which allows to supply a function via options implementing custom comparison of actions. The predicate function, which could be passed until now in the actions argument to perform comparisons, is now no longer supported.

Resolves #6

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -192,13 +193,21 @@ The actions to wait for. It can be either:
...
 
 Type: `Number`   
 Default: 50
 
 The timeout given in milliseconds.
 
+##### matchOrder
+
+Type: `Boolean`
+Default: false
```
Comment (satazor) [inline README.md]: Don't you think that the default value should be `true`?

File: index.js [hunk 2]
```diff
@@ -18,22 +27,49 @@ function waitForActions(store, actions, timeout) {
...
-        actions :
-        (storeActions) => actionsMatch(actions, storeActions);
+    let cancel = () => {};
 
-    // If the store already contains the expected actions, resolve the Promise immediately
-    if (shouldPromiseResolve(store.getActions())) {
-        const promise = Promise.resolve();
+    const promise = new Promise((resolve, reject) => {
+        const actionsMatch = (storeActions) => {
+            if (typeof actions === 'function') {
```
Comment (satazor) [inline index.js]: This part of the diff until line 71 can be simplified a bit, will talk to you personally.

**Review by satazor — CHANGES_REQUESTED** [review]

File: index.js [hunk 3]
```diff
@@ -43,10 +63,9 @@ function waitForActions(store, actions, timeout) {
             reject(new waitForActions.TimeoutError());
         }, timeout);
         const unsubscribe = store.subscribe(() => {
-            if (shouldPromiseResolve(store.getActions())) {
-                teardown();
-                resolve();
-            }
+            const forwardingPromise = settledPromise(store.getActions());
```
Comment (satazor) [inline index.js]:
```
store.subscribe(() => {
     const promise = settledPromise(store.getActions());

     if (promise) {
          resolve(promise);
     }
```

File: index.js [hunk 4]
```diff
@@ -18,22 +27,33 @@ function waitForActions(store, actions, timeout) {
         actions = actions.map((value) => typeof value === 'string' ? { type: value } : value);
     }
 
-    const shouldPromiseResolve = typeof actions === 'function' ?
-        actions :
-        (storeActions) => actionsMatch(actions, storeActions);
+    const settledPromise = (storeActions) => {
```
Comment (satazor) [inline index.js]: Could we move this function to the outer scope and pass in the `actions` as arguments? Maybe name it expectedActions.

File: index.js [hunk 5]
```diff
@@ -3,13 +3,22 @@
 const differenceWith = require('lodash/differenceWith');
 const isMatch = require('lodash/isMatch');
 const isPlainObject = require('lodash/isPlainObject');
+const isEqualWith = require('lodash/isEqualWith');
 
-function actionsMatch(actions, storeActions) {
+function actionsContaining(actions, storeActions) {
```
Comment (satazor) [inline index.js]: actions -> expectedActions.

File: index.js [hunk 6]
```diff
@@ -3,13 +3,22 @@
...
     return differenceWith(actions, storeActions, (action, storeAction) => isMatch(storeAction, action)).length === 0;
 }
 
-function waitForActions(store, actions, timeout) {
-    timeout = timeout || 50;
+function actionsMatchOrder(actions, storeActions) {
```
Comment (satazor) [inline index.js]: actions -> expectedActions.

File: index.js [hunk 7]
```diff
@@ -3,13 +3,22 @@
...
+    return isEqualWith(actions, storeActions, (action, storeAction) => isMatch(storeAction, action));
+}
+
+function waitForActions(store, actions, options) {
```
Comment (satazor) [inline index.js]: actions -> expectedActions.

Comment (satazor) [reply index.js]: Note that we don't need to change the README, I think `expectedActions` describes our intention better for someone that is trying to understand the code.

**Review by satazor — DISMISSED** [review]

File: README.md [hunk 8]
```diff
@@ -139,38 +139,57 @@ await waitForActions(store, [
...
+    { type: 'FETCH_ORDERS_SUCCESS' }
+];
 
-store.dispatch(login('my-username', '123'));
-await waitForActions(store, (storeActions) => {
-	const hasLoginStart = storeActions.some((item) => item.type === 'LOGIN_START' && item.payload.username === 'my-username');
-    const hasFetchOrdersSuccess = storeActions.some((item) => item.type === 'FETCH_ORDERS_SUCCESS');
+store.dispatch(login('my-username', 'my-password'));
+// Matches and LOGIN_START and FETCH_ORDERS_SUCCESS are dispatched
+await waitForActions(store, expectedActions, (expectedActions, storeActions) => {
```
Comment (satazor) [inline README.md]: Isn't the third argument, an options object with a `matcher`?

File: README.md [hunk 9]
```diff
@@ -139,38 +139,57 @@ await waitForActions(store, [
...
+    const hasFetchOrdersSuccess = storeActions.some((action) => action.type === 'FETCH_ORDERS_SUCCESS');
 
     return hasLoginStart && hasFetchOrdersSuccess;
 });
+
+store.clearActions();
+
+store.dispatch(login('my-username', 'wrong-password'));
+// Throws when LOGIN_FAIL is dispatched
+await waitForActions(store, expectedActions, (expectedActions, storeActions) => {
```
Comment (satazor) [inline README.md]: Same here.

File: README.md [hunk 10]
```diff
@@ -139,38 +139,57 @@ await waitForActions(store, [
 ]);
 ```
 
-### Example #3: function (advanced use cases)
+### Example #3: custom action matcher (advanced use cases)
```
Comment (satazor) [inline README.md]: Don't we think that we could simplify the readme a bit now? We could have:

## Installation

`$ npm i... `

## Usage

`waitForActions(store, actions, [options])`

Waits for `actions` to be dispatched on the `store`.
Returns a promise bla bla.

- Example with string actions
- Example with object actions

### Options

- timeout
- matcher

Here we provide info about the matcher implementation, built-in matchers, with examples.

**Review by satazor — DISMISSED** [review]

**Comment by satazor** [comment]
This implementation is much better!

File: README.md (hunk 8 above)
Comment (acostalima) [reply README.md]: Yep. Will fix.

File: README.md (hunk 9 above)
Comment (acostalima) [reply README.md]: Yep. Will fix.

File: README.md [hunk 11]
```diff
@@ -104,7 +104,7 @@ import thunkMiddleware from 'redux-thunk';
 
 const store = configureStore([thunkMiddleware])();
 
-store.dispatch(login('my-username', '12345'));
+store.dispatch(login('my-username', 'my-password'));
 await waitForActions(store, ['LOGIN_START', 'FETCH_ORDERS_SUCCESS']);
```
Comment (satazor) [inline README.md]: Add a new line.

File: README.md [hunk 12]
```diff
@@ -119,10 +119,10 @@ import thunkMiddleware from 'redux-thunk';
 
 const store = configureStore([thunkMiddleware])();
 
-store.dispatch(login('my-username', '123'));
+store.dispatch(login('my-username', 'my-password'));
 // { type: 'LOGIN_START', payload: { username: 'my-username' } }
 // matches
-// { type: 'LOGIN_START', payload: { username, password } }
+// { type: 'LOGIN_START', payload: { username: 'my-username', password } }
```
Comment (satazor) [inline README.md:127]: Add a new line before `waitForActions`.

File: README.md [hunk 13]
```diff
@@ -190,15 +171,70 @@ The actions to wait for. It can be either:
...
+const store = configureStore([thunkMiddleware])();
+const expectedActions = [
+    { type: 'LOGIN_START', payload: { username: 'my-username' } },
+    { type: 'FETCH_ORDERS_SUCCESS' }
+];
+
+store.dispatch(login('my-username', 'my-password'));
+// Throws if LOGIN_FAIL is dispatched or
+// Matches when LOGIN_START and FETCH_ORDERS_SUCCESS are dispatched
+waitForActions(store, expectedActions, { matcher: (expectedActions, storeActions) => {
```
Comment (satazor) [inline README.md]: Add a new line.

File: test/index.spec.js [hunk 14]
```diff
@@ -36,17 +36,17 @@ it('should resolve the promise when a single action is dispatched', async () =>
...
 });
 
-it('should resolve the promise when action creator is dispatched', async () => {
+it('should fulfill the promise when action creator is dispatched', async () => {
     const mockStore = createMockStore([thunkMiddleware]);
     const actions = [action1, action2, action3];
 
     mockStore.dispatch(actionsCreator(actions));
-    await waitForActions(mockStore, actions);
+    await waitForActions(mockStore, [action1, action2, action3]);
```
Comment (satazor) [inline test/index.spec.js]: Add a new line here and in every similar lines.

**Review by satazor — CHANGES_REQUESTED** [review]

File: README.md (hunk 11 above)
Comment (acostalima) [reply README.md]: Done!

File: README.md (hunk 12 above)
Comment (acostalima) [reply README.md:127]: Done!

File: README.md (hunk 13 above)
Comment (acostalima) [reply README.md]: Done!

File: test/index.spec.js (hunk 14 above)
Comment (acostalima) [reply test/index.spec.js]: Done!

**Review by satazor — APPROVED** [review]

**Pull request merged**